certbot-apache/certbot_apache/interfaces.py

```diff
@@ -0,0 +1,281 @@
+"""Parser interfaces."""
+import abc
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ParserNode(object):
+    """
+    ParserNode is the basic building block of the tree of such nodes,
+    representing the structure of the configuration. It is largely meant to keep
+    the structure information intact and idiomatically accessible.
+
+    The root node as well as the child nodes of it should be instances of ParserNode.
+    Nodes keep track of their differences to on-disk representation of configuration
+    by marking modified ParserNodes as dirty to enable partial write-to-disk for
+    different files in the configuration structure.
+
+    While for the most parts the usage and the child types are obvious, "include"-
+    and similar directives are an exception to this rule. This is because of the
+    nature of include directives - which unroll the contents of another file or
+    configuration block to their place. While we could unroll the included nodes
+    to the parent tree, it remains important to keep the context of include nodes
+    separate in order to write back the original configuration as it was.
+    """
+
+    @property
+    @abc.abstractmethod
+    def ancestor(self):
+        """
+        This property contains a reference to ancestor node, or None if the node
+        is the root node of the configuration tree.
+        """
+
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def arguments(self):
+        """
+        This property contains a list of arguments of this ParserNode object.
+        """
+
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def dirty(self):
+        """
+        This property contains a boolean value of the information if this node has
+        been modified since last save (or after the initial parse).
+        """
+
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def save(self, msg):
+        """
+        Save traverses the children, and attempts to write the AST to disk for
+        all the objects that are marked dirty. The actual operation of course
+        depends on the underlying implementation. save() shouldn't be called
+        from the Configurator outside of it's designated save() method in order
```

Comment:
English insanity nit:
```suggestion
        from the Configurator outside of its designated save() method in order
```